changes/gpu-compute-model.tex

```diff
@@ -1,8 +1,93 @@
+GCN3 ISA, the design provides enough flexibility to implement any
+modern GPU architecture that follows a single-instruction, multiple-data
+(SIMD) execution model. This is achieved by following the gem5
+convention of hiding ISA- and architecture-specific functionality behind a
+generic API. By following this convention, the GPU core models
+can be made generic enough to be used across a variety of GPU ISAs, similar
+to the CPU models.
+
+\subsubsection[GPU Thread Context]{GPU Thread Context}
+At a high level the GPU core's execution model can be described as a
+single-instruction, multiple-thread (SIMT) execution model. In
+this execution model threads (i.e., SIMD lanes) are grouped together into
+a \textit{wavefront}
+(using OpenCL terminology, also called a warp in CUDA terminology) and execute the
+same instruction in lock-step on a single SIMD unit.
+
+The wavefront class in gem5 includes all the relevant GPU thread state and
+executes on a single SIMD unit within a compute unit (CU). Multiple wavefronts
+can be combined to form a \textit{work-group} (also called
+a thread block in CUDA terminology) and multiple work-groups can be combined to
+form the GPU kernel. While wavefronts map to single SIMD units, a work-group maps
+to a CU.
+
+\subsubsection[Compute Unit Pipeline]{Compute Unit Pipeline}
+The CU is the primary processing unit in the GPU, which contains many
+CUs. The current model in gem5 supports in-order issue of SIMD and scalar
+instructions processor.
+The workhorse of the compute unit is the SIMD unit, which executes vector
+ALU instructions. The CU model is highly configurable, allowing users to set the
+number of SIMD units per CU as well as the width of each SIMD unit. In addition to
+the SIMD units, each CU can have some arbitrary number of scalar units, scratch
+pad memory, and global memory units (both vector and scalar memory).
+
+The SIMD unit relies on large vector register files (VRF), however the GPU can
+also be configured to include a scalar register file (SRF) for use with scalar
+execution units. Additionally, vector ALU and memory instructions may contain
+scalar operands, and thus vector operations require access to the SRF.
+
+At a high level the CU pipeline modeled in gem5 consists of four stages: \textit{fetch},
+\textit{scoreboard check}, \textit{schedule}, and \textit{execute}. The fetch stage
+is responsible for arbitrating instruction fetch, storing fetched data,
+and decoding instructions for all wavefronts on the CU. The scoreboard check stage is
+responsible for ensuring that intra-wavefront dependences are
+correctly handled before instructions can be scheduled. The schedule stage is
+responsible for arbitrating among the wavefronts presented by
+the scoreboard check stage for each execution unit type. Once an instruction is dispatched
+to an execution unit it is guaranteed to execute.
+This "execute at execute" approach is the same as used by the gem5 in-order and
```

Comment:
```suggestion
This ``execute at execute'' approach is the same as used by the gem5 in-order and
```

As it is, this won't render correctly when compiled.

Reply:
Done